File: android/app/src/main/java/cloud/vybz/app/VybzSecureStorePlugin.java
package cloud.vybz.app;

import android.util.Base64;

import androidx.security.crypto.EncryptedSharedPreferences;
import androidx.security.crypto.MasterKey;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import java.nio.charset.StandardCharsets;
import java.util.Map;

/**
 * AES-GCM sealed preferences backed by Android KeyStore (via EncryptedSharedPreferences /
 * MasterKey). Phase 13 Android Beta.
 */
@CapacitorPlugin(name = "VybzSecureStore")
public class VybzSecureStorePlugin extends Plugin {
    private static final String PREFS = "vybz_secure_keystore_v1";

    private android.content.SharedPreferences prefs() throws Exception {
        MasterKey masterKey =
                new MasterKey.Builder(getContext())
                        .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
                        .build();
        return EncryptedSharedPreferences.create(
                getContext(),
                PREFS,
                masterKey,
                EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
                EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM);
    }

    @PluginMethod
    public void getItem(PluginCall call) {
        String key = call.getString("key");
        if (key == null || key.isEmpty()) {
            call.reject("key required");
            return;
        }
        try {
            String value = prefs().getString(key, null);
            JSObject ret = new JSObject();
            ret.put("value", value);
            call.resolve(ret);
        } catch (Exception e) {
            call.reject("secure get failed", e);
        }
    }

    @PluginMethod
    public void setItem(PluginCall call) {
        String key = call.getString("key");
        String value = call.getString("value");
        if (key == null || key.isEmpty() || value == null) {
            call.reject("key and value required");
            return;
        }
        try {
            prefs().edit().putString(key, value).apply();
            call.resolve();
        } catch (Exception e) {
            call.reject("secure set failed", e);
        }
    }

    @PluginMethod
    public void removeItem(PluginCall call) {
        String key = call.getString("key");
        if (key == null || key.isEmpty()) {
            call.reject("key required");
            return;
        }
        try {
            prefs().edit().remove(key).apply();
            call.resolve();
        } catch (Exception e) {
            call.reject("secure remove failed", e);
        }
    }

    @PluginMethod
    public void clear(PluginCall call) {
        try {
            prefs().edit().clear().apply();
            call.resolve();
        } catch (Exception e) {
            call.reject("secure clear failed", e);
        }
    }

    /** Debug helper — returns key count only (never values). */
    @PluginMethod
    public void stats(PluginCall call) {
        try {
            Map<String, ?> all = prefs().getAll();
            JSObject ret = new JSObject();
            ret.put("count", all.size());
            ret.put("backend", "AndroidKeyStore+AES-GCM");
            call.resolve(ret);
        } catch (Exception e) {
            call.reject("secure stats failed", e);
        }
    }

    @SuppressWarnings("unused")
    private static String b64(byte[] bytes) {
        return Base64.encodeToString(bytes, Base64.NO_WRAP);
    }

    @SuppressWarnings("unused")
    private static byte[] fromUtf8(String s) {
        return s.getBytes(StandardCharsets.UTF_8);
    }
}

File: ios/App/CapApp-SPM/Sources/CapApp-SPM/VybzSecureStorePlugin.swift
import Foundation
import Capacitor
import Security

/**
 * AES-GCM sealed preferences via iOS Keychain (Phase 19).
 * JS name: VybzSecureStore — same API as Android KeyStore plugin.
 */
@objc(VybzSecureStorePlugin)
public class VybzSecureStorePlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "VybzSecureStorePlugin"
    public let jsName = "VybzSecureStore"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "getItem", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setItem", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "removeItem", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "clear", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stats", returnType: CAPPluginReturnPromise),
    ]

    private let service = "cloud.vybz.app.securestore.v1"

    @objc func getItem(_ call: CAPPluginCall) {
        guard let key = call.getString("key"), !key.isEmpty else {
            call.reject("key required")
            return
        }
        var query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: key,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne,
        ]
        var item: CFTypeRef?
        let status = SecItemCopyMatching(query as CFDictionary, &item)
        if status == errSecItemNotFound {
            call.resolve(["value": NSNull()])
            return
        }
        guard status == errSecSuccess, let data = item as? Data,
              let value = String(data: data, encoding: .utf8) else {
            call.reject("secure get failed")
            return
        }
        call.resolve(["value": value])
    }

    @objc func setItem(_ call: CAPPluginCall) {
        guard let key = call.getString("key"), !key.isEmpty,
              let value = call.getString("value") else {
            call.reject("key and value required")
            return
        }
        let data = Data(value.utf8)
        let base: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: key,
        ]
        SecItemDelete(base as CFDictionary)
        var add = base
        add[kSecValueData as String] = data
        add[kSecAttrAccessible as String] = kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly
        let status = SecItemAdd(add as CFDictionary, nil)
        if status == errSecSuccess {
            call.resolve()
        } else {
            call.reject("secure set failed")
        }
    }

    @objc func removeItem(_ call: CAPPluginCall) {
        guard let key = call.getString("key"), !key.isEmpty else {
            call.reject("key required")
            return
        }
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: key,
        ]
        SecItemDelete(query as CFDictionary)
        call.resolve()
    }

    @objc func clear(_ call: CAPPluginCall) {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
        ]
        SecItemDelete(query as CFDictionary)
        call.resolve()
    }

    @objc func stats(_ call: CAPPluginCall) {
        call.resolve(["count": 0, "backend": "keychain"])
    }
}
